### ISSUE TITLE
cmd/go: vet incorrectly flags loop variable captured by func literal in GOPATH mode

### ISSUE THREAD

**Issue by rittneje** [golang/go#70641]
### Go version

go version go1.22.9 darwin/amd64

### Output of `go env` in your module/workspace:

```shell
n/a
```

### What did you do?

I wrote the follow unit test in GOPATH mode.

```go
package looptest

import "testing"

func TestLoopSemantics(t *testing.T) {
	values := []string{"1", "2"}
	for _, x := range values {
		t.Run(x, func(t *testing.T) {
			t.Parallel()
			t.Log(&x)
		})
	}
}
```

Then I ran `go vet` on the package.

### What did you see happen?

```
$ go vet looptest
# looptest
# [looptest]
src/looptest/loop_test.go:10:11: loop variable x captured by func literal
```

### What did you expect to see?

No issues.

In Go 1.22+, the compiler uses the new loop semantics, even in GOPATH mode. This is proven by running the test, which gives output like so. (The exact address is different each time of course, but the important part is that `x` has a different address in each test case.)

```go
$ go test looptest -v
=== RUN   TestLoopSemantics
=== RUN   TestLoopSemantics/1
=== PAUSE TestLoopSemantics/1
=== RUN   TestLoopSemantics/2
=== PAUSE TestLoopSemantics/2
=== CONT  TestLoopSemantics/1
=== CONT  TestLoopSemantics/2
=== NAME  TestLoopSemantics/1
    loop_test.go:10: 0xc000032350
=== NAME  TestLoopSemantics/2
    loop_test.go:10: 0xc0000323d0
--- PASS: TestLoopSemantics (0.00s)
    --- PASS: TestLoopSemantics/1 (0.00s)
    --- PASS: TestLoopSemantics/2 (0.00s)
PASS
ok      looptest        0.498s
```

Therefore, `go vet` should not have applied this check at all.

**Comment by mknyszek** [comment]
CC @golang/tools-team

**Comment by findleyr** [comment]
CC @adonovan @timothy-king 

It sounds like the loopvar analyzer needs to be made aware of this behavior in GOPATH mode.

@matloob can you clarify the semantics here? Should we assume the effective language version in GOPATH mode is the runtime version?

**Comment by ianlancetaylor** [comment]
My understanding is that in the absence of a go.mod file the go tool uses Go 1.16 semantics.  So the warning is correct when not using modules.

**Comment by rsc** [comment]
@ianlancetaylor I think we intended to limit GOPATH mode to Go 1.16 at first, but when we decided to keep GOPATH mode around for people who are for one reason or another not using modules, we decided to let the language version float forward, since otherwise you couldn't use generics ever, for example. The behavior of the 'go test' confirms that. Or at least, 'go test' and 'go vet' clearly disagree here.

**Comment by ianlancetaylor** [comment]
Ah, thanks.

**Comment by matloob** [comment]
This was previously reported as #65612 in the command-line-arguments context, but it's the same underlying issue for GOPATH. We fixed this in 1.23 with https://go.dev/cl/593156, explicitly setting the go version to the toolchain version when outside of a module.

We considered backporting the change to Go 1.22 but decided not to because it didn't seem to be a serious problem that prevented a program from working at all (and Go 1.22 already had issues with not passing in -lang to the compiler that we wanted to fix at the same time, and would change the behavior for building programs).

**Comment by findleyr** [comment]
Thanks @matloob, I'd missed (or forgotten) that context.

I think it's reasonable not to backport this fix for 1.22 GOPATH mode. Therefore, closing as a dupe.
